src/session/session_op_tracker.c

```diff
@@ -0,0 +1,252 @@
+		return (0);
+
+	/* Don't capture tracing unless we are recording for the API */
+	if (!api_boundary && TAILQ_EMPTY(&session->op_trackerq))
+		return (0);
+
+	/*
+	 * Don't capture WT_CONNECTION methods. It would be nice to have
+	 * a different way to capture this. An alternative would be to have
+	 * an exclude list built via op_track.py. In the mean time this
+	 * works.
+	 */
+	if (type >= WT_OP_TYPE_WT_CONNECTION_async_flush &&
+	    type <= WT_OP_TYPE_WT_CONNECTION_get_extension_api)
+		return (0);
+
+	WT_ASSERT(session, session->api_call_depth == 0 ||
+	    !TAILQ_EMPTY(&session->op_trackerq));
+
+	/*
+	 * If we are entering the first time via a public API clear any
+	 * tracked operations from the previous API call. Track the call depth
+	 * since WiredTiger uses API calls internally.
+	 */
+	if (api_boundary && session->api_call_depth == 0)
+		WT_RET(__wt_session_op_tracker_clear(session));
+
+	if ((entry = TAILQ_FIRST(&session->op_tracker_availq)) != NULL) {
+		TAILQ_REMOVE(&session->op_tracker_availq, entry, aq);
+	} else
+		WT_RET(__wt_calloc_one(session, &entry));
+
+	WT_RET(__wt_epoch(session, &entry->start));
+	memcpy(&entry->last_stop, &entry->start, sizeof(struct timespec));
+	entry->api_boundary = api_boundary;
+	TAILQ_INSERT_TAIL(&session->op_trackerq, entry, q);
+
+	/*
+	 * Find the parent for this entry. It is the last entry in the queue
+	 * that hasn't been finished.
+	 */
+	for (prev_entry = TAILQ_PREV(entry, __op_tracker, q);
+	    prev_entry != NULL;
+	    prev_entry = TAILQ_PREV(prev_entry, __op_tracker, q)) {
+		if (prev_entry->done)
+			continue;
+		prev_entry->self_time_us +=
+		    WT_TIMEDIFF(prev_entry->last_stop, entry->start);
+		break;
+	}
+
+	/* Create and insert the first entry into the slow op tracker. */
+	entry->type = type;
+	*entryp = entry;
+
+	if (api_boundary)
+		++session->api_call_depth;
+
+	WT_ASSERT(session,
+	    !TAILQ_EMPTY(&session->op_trackerq));
+
+	return (0);
+}
+/*
+ * __wt_session_op_tracker_finish_entry --
+ *	Finalize an entry.
+ */
+int
+__wt_session_op_tracker_finish_entry(
+    WT_SESSION_IMPL *session, WT_OP_TRACKER_ENTRY *entry)
+{
+	WT_OP_TRACKER_ENTRY *prev_entry;
+
+	/*
+	 * Don't track operations completed by internal sessions. It's also
+	 * OK to ignore calls with a NULL entry - they are operations that
+	 * weren't fully created, and we handle finish with a NULL entry to
+	 * avoid complicating callers code.
+	 */
+	if (F_ISSET(session, WT_SESSION_INTERNAL) || entry == NULL)
+		return (0);
+
+	/*
+	 * Special case for connection close - skip handles that are no longer
+	 * valid. It would be nice to capture this elsewhere, but automatically
+	 * capturing all API calls makes that non-trivial.
+	 */
+	if (session->iface.connection == NULL)
+		return (0);
+
+	if (entry->api_boundary)
+		--session->api_call_depth;
+
+	WT_RET(__wt_epoch(session, &entry->end));
+	entry->done = 1;
+	/* Start the sub-entry timer if this is a nested operation. */
```

Comment:
To be clear, this is the sub-entry timer within the "parent" object. Not the timer object of a sub-entry